### PR TITLE
Simplify regexes

### PR DESCRIPTION
**Pull request by kinow**
When using `[...]` in regexes, I believe we do not need to escape `.`, `:`, `)`, `<`, `>`, etc. As far as I know, we need to escape only the closing `]`, and the normal sequences like `\n`, `\r`, `\t`.

Python regex engine does the job of escaping these character anyway, but being regexes not always the fastest option, perhaps removing unnecessary escapes is a good idea?

Also included one more change (second commit) to use ```([`\])``` instead of ```(`|\)```. For two characters it makes no difference performance-wise. With more characters ```[`\]``` would probably be faster. But it's still easier to read I think?

Low priority, for Cylc 8 or later, without backport to Cylc 7 (I do not think it's necessary for the 7.8.x release series).

Cheers
Bruno

### CODE REVIEW

**Comment by kinow** [comment]
Waited for Travis-CI to build the branch in my repository. If it fails here, it's one of those flaky tests, and Travis will need a kick.

**Comment by matthewrmshin** [comment]
Yes. For completeness, the only other character that needs escaping in `[...]` is the hyphen/minus sign `-`, which is used in syntax like `[a-z]`.

The only other thing we can do is to improve performance of regular expression logic is to pre-compile them where possible. (Not sure if this is still true or not.)

This change looks like good to go.

File: lib/cylc/param_expand.py [hunk 1]
```diff
@@ -64,15 +64,15 @@ def expand(template, params, results, values=None):
 from parsec.OrderedDict import OrderedDictWithDefaults
 
 # To split runtime heading name lists.
-REC_NAMES = re.compile(r'(?:[^,<]|\<[^>]*\>)+')
+REC_NAMES = re.compile(r'(?:[^,<]|<[^>]*>)+')
```
Comment (oliver-sanders) [inline lib/cylc/param_expand.py]: I think this one may be dangerous:

```console
$ sed -r 's/<foo>/x/' <<<'a<foo>b
axb
$ sed -r 's/\<foo\>/x/' <<<'a<foo>b'
a<x>b
```

Comment (matthewrmshin) [reply lib/cylc/param_expand.py]: Can't remember if Python's `re` support `\<` and `\>` for word boundaries or not. I normally use `\b` for word boundary.

Comment (oliver-sanders) [reply lib/cylc/param_expand.py]: Ditto.

Comment (oliver-sanders) [reply lib/cylc/param_expand.py]:
```console
$ python -c "import re; print(re.sub(r'\<fun\>', r'a fun b', 'x'))"
x
```

Comment (hjoliver) [reply lib/cylc/param_expand.py]:
```
re.sub(pattern, repl, string, count=0, flags=0)
```
Args 'repl' and 'string' switched above? 

```
python -c "import re; print(re.sub(r'\<fun\>', 'x', r'a fun b'))"
a fun b

python -c "import re; print(re.sub(r'\bfun\b', 'x', r'a fun b'))"
a x b
```

Comment (hjoliver) [reply lib/cylc/param_expand.py]: Anyhow,

> I think this one may be dangerous:

Not dangerous, right?

Comment (oliver-sanders) [reply lib/cylc/param_expand.py]: > Can't remember if Python's re support \< and \> for word boundaries or not

Apparently not!

```python
import re

FOO = '''
    foo => <bar> => baz
    foo<bar> => baz
    foo => baz<bar>
'''

print('before')
REC_NAMES = re.compile(r'\<[^>]*\>')
for match in REC_NAMES.finditer(FOO):
    print(match)


print('\nafter')
REC_NAMES = re.compile(r'<[^>]*>')
for match in REC_NAMES.finditer(FOO):
    print(match)

print('\nword break')
REC_NAMES = re.compile(r'\b[^>]*\b')
for match in REC_NAMES.finditer(FOO):
    print(match)
```

```
before
<re.Match object; span=(12, 17), match='<bar>'>
<re.Match object; span=(32, 37), match='<bar>'>
<re.Match object; span=(59, 64), match='<bar>'>

after
<re.Match object; span=(12, 17), match='<bar>'>
<re.Match object; span=(32, 37), match='<bar>'>
<re.Match object; span=(59, 64), match='<bar>'>

word break
<re.Match object; span=(5, 8), match='foo'>
<re.Match object; span=(8, 8), match=''>
<re.Match object; span=(13, 16), match='bar'>
<re.Match object; span=(16, 16), match=''>
<re.Match object; span=(21, 36), match='baz\n    foo<bar'>
<re.Match object; span=(36, 36), match=''>
<re.Match object; span=(41, 52), match='baz\n    foo'>
<re.Match object; span=(52, 52), match=''>
<re.Match object; span=(56, 63), match='baz<bar'>
<re.Match object; span=(63, 63), match=''>

```

> Not dangerous, right?

Seems this is safe in Python.

**Pull request merged**